flink-staging/flink-scala-shell/src/main/java/org/apache/flink/api/java/ScalaShellRemoteStreamEnvironment.java

```diff
@@ -0,0 +1,155 @@
+	private static final Logger LOG = LoggerFactory.getLogger(ScalaShellRemoteStreamEnvironment.class);
+	// reference to Scala Shell, for access to virtual directory
+	private FlinkILoop flinkILoop;
+	/**
+	 * Creates a new RemoteStreamEnvironment that points to the master
+	 * (JobManager) described by the given host name and port.
+	 *
+	 * @param host	 The host name or address of the master (JobManager), where the
+	 *				 program should be executed.
+	 * @param port	 The port of the master (JobManager), where the program should
+	 *				 be executed.
+	 * @param jarFiles The JAR files with code that needs to be shipped to the
+	 *				 cluster. If the program uses user-defined functions,
+	 *				 user-defined input formats, or any libraries, those must be
+	 */
+	public ScalaShellRemoteStreamEnvironment(String host, int port, FlinkILoop flinkILoop, String... jarFiles) {
+		super(host, port, jarFiles);
+		this.flinkILoop = flinkILoop;
+	}
+	/**
+	 * compiles jars from files in the shell virtual directory on the fly, sends and executes it in the remote stream environment
+	 *
+	 * @return Result of the computation
+	 * @throws ProgramInvocationException
+	 */
+	@Override
+	public JobExecutionResult execute() throws Exception {
+		prepareJars();
+		return(super.execute());
+	}
+	/**
+	 * prepares the user generated code from the shell to be shipped to JobManager
+	 * (i.e. save it into jarFiles of this object)
+	 */
+	private void prepareJars() throws MalformedURLException {
+		String jarFile = flinkILoop.writeFilesToDisk().getAbsolutePath();
+
+		// get "external jars, and add the shell command jar, pass to executor
+		List<String> alljars = new ArrayList<String>();
+		// get external (library) jars
+		String[] extJars = this.flinkILoop.getExternalJars();
+
+		if(!ArrayUtils.isEmpty(extJars)) {
+			alljars.addAll(Arrays.asList(extJars));
+		}
+		// add shell commands
+		alljars.add(jarFile);
+		String[] alljarsArr = new String[alljars.size()];
+		alljarsArr = alljars.toArray(alljarsArr);
+		for (String jarF : alljarsArr) {
+			URL fileUrl = new URL("file://" + jarF);
+			System.out.println("sending:" + fileUrl);
+			try {
+				JobWithJars.checkJarFile(fileUrl);
+			}
+			catch (IOException e) {
+				throw new RuntimeException("Problem with jar file " + fileUrl, e);
+			}
+			jarFiles.add(fileUrl);
+		}
+	}
+	/**
+	 * compiles jars from files in the shell virtual directory on the fly, sends and executes it in the remote stream environment
+	 * @param jobName name of the job as string
+	 * @return Result of the computation
+	 * @throws ProgramInvocationException
+	 */
+	@Override
+	public JobExecutionResult execute(String jobName) throws ProgramInvocationException {
+		try {
+			prepareJars();
+		} catch (MalformedURLException e) {
+			e.printStackTrace();
+		}
+
+		JobExecutionResult jer = null;
+		try {
+			jer = super.execute(jobName);
+		} catch (Exception e) {
+			e.printStackTrace();
```

Comment:
Removed try catch block, analogous to the batch case it will be thrown at the user directly now.